Instructions + scripts for running tests repeatedly (#6115)
microsoft/MixedRealityToolkit-Unity: Documentation/Contributing/UnitTests.md

# Writing and Running Tests in MRTK
MRTK has a set of tests to ensure that changes to our code do not regress existing behavior. Before you submit a pull request, make sure to :

1. Run the tests locally to make sure your changes don't regress existing behavior (you will not be able to check in if any tests fail)

2. If fixing a bug, write a test to prevent others from breaking your bugfix.

2. If writing a feature, write new tests to ensure that other people don't break your feature in the future.

## Running tests
### Running tests from Unity editor
The [Unity Test Runner](https://docs.unity3d.com/Manual/testing-editortestsrunner.html) can be found under Window > General > Test Runner and will show all available MRTK play and edit mode tests. 

### Running tests from command line

> You can also run the [powershell](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell?view=powershell-6) script located at `Scripts\test\run_playmode_tests.ps1`. This will run the playmode tests exactly as they are executed on github / CI (see below), and print results. Here are some examples of how to run the script

Author: Oh, I didn't know about that! I never actually tested that experience. I suppose you could just run that command line tests multiple times and parse output. I think the approach of modifying the code is still faster since every time you run a test from cmd unity does a bunch of extra work. I can mention the command line arguments though.

> You can also run the [powershell](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell?view=powershell-6) script located at `Scripts\test\run_playmode_tests.ps1`. This will run the playmode tests exactly as they are executed on github / CI (see below), and print results. Here are some examples of how to run the script

Reviewer: Exactly they have different benefits. The python script is great to target one test case and get multiple results in editor but you have to run and copy and paste and then delete/cleanup

Where a command line of run everything 5-10 times while I go to lunch involves less setup and covers more bases for that need

> You can also run the [powershell](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell?view=powershell-6) script located at `Scripts\test\run_playmode_tests.ps1`. This will run the playmode tests exactly as they are executed on github / CI (see below), and print results. Here are some examples of how to run the script

Author: Done, see run_repeat_tests.ps1

    .\run_playmode_tests.ps1 H:\mrtk.dev -unityExePath = "C:\Program Files\Unity\Hub\Editor\2018.4.1f1\Editor\Unity.exe" -outFolder "C:\playmode_test_out\"
```

### Running tests via pull request validation
MRTK's CI will build MRTK in all configurations and run all edit and play mode tests. CI can be triggered by posting a comment on the github PR `/azp run mrtk_pr` if the user has sufficient rights. CI runs can be seen in the 'checks' tab of the PR. 

Only after all of the tests passed successfully the PR can be merged into mrtk_development. 

### Running stress tests / bulk tests
Sometimes tests only fail occasionally, it can be frustrating to debug these failures. The best way we have found to reliably repeat tests is to modify scripts locally, since unity's test system [does not currently support the Retry attribute](https://docs.unity3d.com/Packages/com.unity.test-framework@1.0/manual/index.html).

To run multiple tests locally, modify your test to call your test multiple times. We have provided a python script to make this more convenient.

First, make sure you have [Python 3.X installed](https://www.python.org/downloads/).

> Let's say you have a test that you want to run many times:

Reviewer 2: ```suggestion
For a single test that needs to me executed multiple times:
```

```
[UnityTest]
public IEnumerator MyTest() {...}
```

> Run the following from a command line (we recommend [PowerShell](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell?view=powershell-6#powershell-core))

Reviewer 2: ```suggestion
Run the following from a command line ([PowerShell](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell?view=powershell-6#powershell-core) is recommended)
```

```
cd scripts\tests
# Repeat the test 5 times. Default is 100 
python .\generate_repeat_tests.py -n 5 -t MyTest
```

> Copy and paste the output into your test file. If you need to run multiple tests in sequence:

Reviewer 2: ```suggestion
Copy and paste the output into your test file. The following script is for running multiple tests in sequence:
```

```
cd scripts\tests
# Repeat the test 5 times. Default is 100 
python .\generate_repeat_tests.py -n 5 -t MyTest MySecondTest
```

> Your new test file should now contain

Reviewer 2: ```suggestion
The new test file should now contain
```

```
[UnityTest]
public IEnumerator A1MyTest0(){ yield return MyTest();}
[UnityTest]
public IEnumerator A2MyTest0(){ yield return MyTest();}
[UnityTest]
public IEnumerator A3MyTest0(){ yield return MyTest();}
[UnityTest]
public IEnumerator A4MyTest0(){ yield return MyTest();}
[UnityTest]
public IEnumerator MyTest() {...}
```

> Open the test runner and you should see your new tests that you can call repeatedly.

Reviewer 2: ```suggestion
Open the test runner and observe the new tests that can now be called repeatedly.
```

> ## Writing Tests

Reviewer 2: ```suggestion
## Writing tests
```

> To ensure MRTK being a stable and reliable toolkit, every feature should come with unit tests and sample usage in one of the example scenes. Having good test coverage in a big codebase like MRTK is crucial for stability and having confidence when doing changes in code.

Reviewer 2: this is explaining our test setup - i would move that to the top of the page bc it's prerequisite for going through your instructions (by then you should already know that we're using unity test runner)
